### PR TITLE
Broadcasting + better CuArrays support

### PR DESCRIPTION
**Pull request by colinxs**
1. Defines `Base.similar(A::ElasticArray, ::Type{T}, dims::Dims{N})` to return an ElasticArray.
2. Changes broadcasting to return an `ElasticArray`
3. Removes `Base.similar(::Type{ElasticArray{T}}, dims::Dims{N})`; the definition for `AbstractArray` in `Base` is sufficient.
4. Removes `parent(A::ElasticArray)`. Several functions in `LinearAlgebra` depend on `size(A) == size(parent)`. Given that `ElasticArray <: DenseArray` and we define `Base.pointer` this doesn't affect the ability to hit `BLAS`. Not sure what the other ramifications would be. The other alternative that works but feels wrong is `parent(A::ElasticArray) = reshape(A.data, size(A))` but that seems worse?
5. Adds `Adapt.adapt_structure(to, A::ElasticArray)` so the `CuArrays.cu(A::ElasticArray)` returns an `ElasticArray` wrapping a `CuArray`.

For `A::ElasticArray`, `A * A`, `2 * A`, and `inv(A)` now return an `ElasticArray`. The goal is for matrix operations to always return an `ElasticArray`, but without more test coverage I'm no sure if I'm missing something.

This depends on the changes in #20 so we can wait until that's merged. I also need to add a test for `adapt_structure`. I still hit scalar indexing with CuArrays, so need to figure that out as well.

### CODE REVIEW

File: src/elasticarray.jl [hunk 1]
```diff
@@ -57,76 +67,139 @@ ElasticArray(A::AbstractArray{T,N}) where {T,N} = copyto!(ElasticArray{T,N}(unde
 Base.convert(::Type{T}, A::AbstractArray) where {T<:ElasticArray} = A isa T ? A : T(A)
 
 
+@inline function Base.similar(A::ElasticArray, ::Type{T}, dims::Dims{N}) where {T,N}
```
Comment (oschulz) [inline src/elasticarray.jl:70]: I had this, in the past, but changed it due to a request by @garrison,  see #5. I'm not firmly against it, we should check the implications and pull @garrison into the discussion, if he's interested.

Is defining `Base.similar` to return an `ElasticArray` important for the rest of this PR? If not, I suggest to make a separate PR for it, so it can be discussed in a more narrow scope.

Comment (oschulz) [reply src/elasticarray.jl:70]: Maybe we could use `_similar_ea()` for now, within this PR, and then have the separate PR for `Base.similar`?

File: src/elasticarray.jl [hunk 2]
```diff
@@ -57,76 +67,139 @@ ElasticArray(A::AbstractArray{T,N}) where {T,N} = copyto!(ElasticArray{T,N}(unde
...
+@inline Base.length(A::ElasticArray) = length(A.data)
 
-import Base.==
-(==)(A::ElasticArray, B::ElasticArray) =
-    ndims(A) == ndims(B) && A.kernel_size == B.kernel_size && A.data == B.data
+@propagate_inbounds Base.getindex(A::ElasticArray, i::Int) = getindex(A.data, i)
 
+@propagate_inbounds Base.setindex!(A::ElasticArray, x, i::Int) = setindex!(A.data, x, i)
 
-Base.parent(A::ElasticArray) = A.data
```
Comment (oschulz) [inline src/elasticarray.jl]: Why remove `Base.parent`? Or did it just move somewhere else (don't see it right now)?

**Comment by oschulz** [comment]
Could you rebase this to the current master? Will make review easier.

**Comment by oschulz** [comment]
Bump @colinxs  :-)

**Comment by colinxs** [comment]
Thanks for the reminder @oschulz! I rebased it.

I'll try to make that first `FlattenedArray` PR in the next few days.

**Review by oschulz — CHANGES_REQUESTED** [review]

**Comment by oschulz** [comment]
Ok, done with the review - the only tricky thing is `Base.similar(A::ElasticArray, ...)`, see my comments.

File: src/elasticarray.jl (hunk 1 above)
Comment (colinxs) [reply src/elasticarray.jl:70]: To my knowledge, #6 addresses two separate issues. The first is supporting BLAS via `Base.unsafe_convert` while the second deals with whether `Base.similar(A::ElasticArrays, ...)` should return an `ElasticArray` or something else.

For proper CuArrays support, we do need to override `Base.similar(A::AbstractArray, ...)`, which is called in several places in e.g. `LinearAlgebra`. Without it, you'd get the following behavior:
```julia
julia> using ElasticArrays, CuArrays

julia> A = cu(ElasticArray{Float32}(undef, 2, 2));

julia> B = cu(ElasticArray{Float32}(undef, 2, 2));

julia> typeof(A * B)
Array{Float32,2}
```
That is, `A` and `B` are on the GPU but `A * B` is on the CPU, which is both slow and inconsistent with CuArrays:
```julia
julia> typeof(cu(rand(2,2)) * cu(rand(2,2)))
CuArray{Float32,2,Nothing}
```

That being said, we don't *have* to return an `ElasticArray`. The following would work:
```julia
Base.similar(A::ElasticArray, T::Type, dims::Dims) = similar(A.data, T, dims)
```

As for whether we *should* return an `ElasticArray` or not, I would argue that if I have `A::ElasticArray` and `B::ElasticArray` that it's reasonable to expect that `A * B` is also an `ElasticArray` (and is also consistent with e.g. StaticArrays/CuArrays etc.). But that also may require more effort on our part :).

File: src/elasticarray.jl (hunk 2 above)
Comment (colinxs) [reply src/elasticarray.jl]: Because at least one stdlib function depends on `size(A::AbstractArray) == size(parent(similar(A, ...)))`. Here's a snippet from `LinearAlgebra.inv`:
```julia
Ai = inv!(lu(AA))
Ai = convert(typeof(parent(Ai)), Ai)
```

Before adding `Base.similar(A::ElasticArray, ...)`, `inv!(lu(AA))` returned an `Matrix`, but now it returns an `ElasticArray`. So when we do `convert(typeof(parent(Ai)), Ai)` we try to convert a matrix into a vector (because `typeof(parent(Ai)) <: AbstractVector`), which fails.

If we change `Base.similar` to `Base.similar(A::ElasticArray, T::Type, dims::Dims) = similar(A.data, T, dims)` as mentioned above then we can add back in `Base.parent`.

But from the docs for `Base.parent`
> Returns the "parent array" of an array view type (e.g., SubArray), or the array
  itself if it is not a view.

Is `ElasticArray` really a "view type"? Do we need `Base.parent`? I'm unsure :)

File: src/elasticarray.jl (hunk 1 above)
Comment (oschulz) [reply src/elasticarray.jl:70]: > As for whether we should return an ElasticArray [...] reasonable to expect that A * B is also an ElasticArray

That was my initial opinion, too, but @garrison argued otherwise (as to what to expect from `similar`) in #6.

But all that was a while ago - @garrison, do you want to weigh in here?

Comment (oschulz) [reply src/elasticarray.jl:70]: Hm, if @garrison has no strong opinion here I suggest we merge this.

**Comment by oschulz** [comment]
Thanks again for this, @colinxs!

Can you battle test it in your application(s) for a few day, before we make a release? Or do have have further changes that you'd like to make first?

**Comment by garrison** [comment]
Just saw the discussion (sorry I did not reply sooner).  I will test `master` with all my existing code and let you know if anything breaks.

**Comment by colinxs** [comment]
@oschulz yup! No worries. My only remaining concerns are:

1) Add a test for `adapt_structure`.
2) I was hitting scalar indexing when the parent was a `CuArray`, so we'll need to look into how to properly wrap a CuArray. Perhaps @aterenin could take a look at this as well since he's earlier PR suggested he was ElasticArrays + CuArrays.
3) We test that for `A::ElasticArray`, `A * A`, `2 * A`, and `inv(A)` return an `ElasticArray`, but we might be missing something.
4) We should probably do a benchmark comparison between this PR and an `Array` for array math and broadcasting to make sure this doesn't slow anything down.

**Comment by aterenin** [comment]
My PR was very tiny and made in order to allow compatibility, and make things like this feasible. To fix scalar getindex, I think one would have to implement something that tells the broadcasting machinery that it should promote broadcasting to the child array. I don't know the internals for broadcasting well enough to know how to do this.

**Comment by oschulz** [comment]
@garrison , thanks, I'll wait for your feedback before making a release.

**Comment by oschulz** [comment]
@aterenin, sorry, which PR?

**Comment by oschulz** [comment]
> @colinxs: My only remaining concerns are [...]

Sure, we don't need to rush a release, please do let me know how the benchmarks turn out. I have a feeling we're all aiming for top performance here.

**Comment by garrison** [comment]
Happy to report that all of my code works as expected with `master`.

One thing I did notice a few weeks ago is that with #20, it is now possible to `append!` an array whose `size` is incompatible with the `ElasticArray`, as long as the `length` of the iterable agrees.  I'm a bit hesitant about this, but also welcome that ElasticArrays will now work with appending arbitrary iterables rather than arrays.  Perhaps an additional check should be done, though, if it is something that has a definite `size`?

**Comment by garrison** [comment]
Also, if there is to be any hope of recovering after an exception, shouldn't `_check_size` be called before actually modifying any data?

**Comment by oschulz** [comment]
> One thing I did notice a few weeks ago is that with #20, it is now possible to append! an array whose size is incompatible with the ElasticArray, as long as the length of the iterable agrees.

You mean if the shape is wrong, but the length compatible? If so, I'd say that since 

```julia
append!(rand(6), rand(3,4))
```

is allowed

```julia
append!(ElasticArray{Float64}(undef, 3, 4), rand(6))
```

should be allowed too.

**Comment by oschulz** [comment]
> Also, if there is to be any hope of recovering after an exception, shouldn't _check_size be called before actually modifying any data

Oh, yes, that would probably make sense! :-)

@colinxs ?

**Comment by oschulz** [comment]
Thanks for helping out with testing, @garrison !

**Comment by garrison** [comment]
> If so, I'd say that since

This is a good point.  But one could also make the case that `append!(rand(6), rand(3,4))` is different conceptually from `append!(rand(3,4), rand(6))`, and the latter is more like the example you gave.  My concern here is that somebody will call something like `ElasticArray{Float64}(undef, 3, 4), rand(4, 3))` and not get the expected result/data alignment.

**Comment by oschulz** [comment]
Hm, yes ... maybe always allow to append a vector or iterator, but higher-dimensional arrays only when compatible? @colinxs , what do you think?

**Comment by colinxs** [comment]
> Also, if there is to be any hope of recovering after an exception, shouldn't _check_size be called before actually modifying any data

I thought about doing that, but that requires either:
1) Iterating through `iter` twice, first the check the sizes, the second to append the data (2N iterations + 1 `resize!` + 1 `copyto!`)
2) Appending each element one-by-one (N iterations + N `resize!`s + N `copyto!`s)
It may not be possible to iterate over `iter` twice.

If you look at e.g. `Base.copyto!` for an `AbstractArray` it may fail partway through, but I agree this is a bit different since the `ElasticArray` may end up in a corrupted state. So perhaps the best middle ground is append `iter` in batch if `iter::AbstractArray` (as we can check the size beforehand), otherwise do (2).

> This is a good point. But one could also make the case that append!(rand(6), rand(3,4)) is different  ...

Maybe, but that starts to treat `ElasticArray` as an array-of-arrays, which its not. As for user's making errors, god forbid they do `append!(ElasticArray{Float64}(undef, 3, 3), rand(3,3)`. No size checking helps you there haha. Both sides have good points, but I'm partial to allowing `append!(ElasticArray{Float64}(undef, 3, 4), rand(4,3)`.

cc : @garrison @oschulz

**Comment by colinxs** [comment]
> Appending each element one-by-one (N iterations + N resize!s + N copyto!s)
It may not be possible to iterate over iter twice.

I just realized that you could optimize this as well by using `HasLength`/`HasEltype` and re-creating the machinery in `Base._append!`

**Comment by oschulz** [comment]
Hm, what about a try/catch, and we truncate/revert and rethrow if necessary?

**Comment by colinxs** [comment]
Seems kind of hacky (and try/catch has it's own performance issues). I think our best option is to just recreate the machinery in Base.

**Comment by colinxs** [comment]
@oschulz @garrison 

Boom done: https://github.com/JuliaArrays/ElasticArrays.jl/pull/23

**Comment by colinxs** [comment]
Also, in reference to the discussion whether `append!(ElasticArray{Float64}(undef, 3, 4), rand(4,3))` should be allowed or not, it's _always_ been possible with ElasticArrays, #20 didn't change that. Here's the previous definition: 
```
function Base.append!(dest::ElasticArray, src::AbstractArray)
    rem(length(eachindex(src)), dest.kernel_length) != 0 && throw(DimensionMismatch("Can't append, length of source array is incompatible"))
    append!(dest.data, src)
    dest
end
```
So it only checked that `length(src)==dest.kernel_length` and not `front(size(src)) == dest.kernel_size`.

With that in mind I think we should go ahead and tag a release that includes #20 and #23. I don't have time right now to address the points in https://github.com/JuliaArrays/ElasticArrays.jl/pull/21#issuecomment-624410618, so if we don't feel comfortable with that right now you can go ahead and revert f468be5680a50dec9a878a70f0c485f1ded42368 and re-open this PR for a later merge.

**Comment by oschulz** [comment]
> it's always been possible with ElasticArrays, #20 didn't change that.

Yes, I think we should discuss this in a separate issue, it shouldn't block the release. @garrison, since you brought it up, do you want to create an issue on append and shape of the "appendee"?

**Comment by oschulz** [comment]
Ok, all merged - I would call the result v1.1, I'd say the changes we not really breaking. Or would you prefer a v2, @colinxs and @garrison?

**Comment by garrison** [comment]
> @garrison, since you brought it up, do you want to create an issue on append and shape of the "appendee"?

Sure.



> I would call the result v1.1, I'd say the changes we not really breaking.

Sounds good to me.

**Comment by oschulz** [comment]
Ok, then I'll do a few more application tests myself over the weekend, and then tag a release early next week.

**Comment by colinxs** [comment]
Will this next release include https://github.com/JuliaArrays/ElasticArrays.jl/commit/f468be5680a50dec9a878a70f0c485f1ded42368? Or do you plan to revert it temporarily? I think either option is fine, just curious.

As for the next version, _technically_ it should be v2.0.0 because the addition of the type parameter `V` is a breaking change.

**Comment by oschulz** [comment]
> Will this next release include f468be5?

My plan is to base the release on the current master, no reverts.

**Comment by colinxs** [comment]
Okay, sounds good! Any chance you'd be able to tag a release _today_? Haha I'm waiting to make a few releases that depend on #20 and #23. For what's it's worth, the fork of ElasticArrays that I've been using for a few weeks includes those changes, so it's fairly well tested.

**Comment by colinxs** [comment]
I'll be on hand to patch any bugs that may come up :)

**Comment by oschulz** [comment]
> I'll be on hand to patch any bugs that may come up :)

Now that's an offer! :-)

Here we go: JuliaRegistries/General#14361

**Comment by colinxs** [comment]
Awesome! Thank you :)

**Comment by oschulz** [comment]
You're very welcome - thanks for contributing!
